ops/framework.py

```diff
@@ -194,6 +194,9 @@ def __init__(self, parent, key):
         kind = self.handle_kind
         if isinstance(parent, Framework):
             self.framework = parent
+            # Avoid Framework instances having a circular reference to themselves.
+            if self.framework is self:
+                self.framework = weakref.proxy(self.framework)
```

Comment:
This isn't strictly required for this PR, but it came up during testing and seems like something we should have.  Could also make it unconditional, if that seems better.